arrow-libs/core/arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt

```diff
@@ -550,8 +429,185 @@ object NonEmptyListSemigroup : Semigroup<NonEmptyList<Any?>> {
+  map: (A, B, C, D) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit
+  ) { aa, bb, cc, dd, _, _, _, _, _, _ ->
+    map(aa, bb, cc, dd)
+  }
+
+inline fun <A, B, C, D, E, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  map: (A, B, C, D, E) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d, e,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit
+  ) { aa, bb, cc, dd, ee, _, _, _, _, _ ->
+    map(aa, bb, cc, dd, ee)
+  }
+
+inline fun <A, B, C, D, E, F, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  f: NonEmptyList<F>,
+  map: (A, B, C, D, E, F) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d, e, f,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit
+  ) { aa, bb, cc, dd, ee, ff, _, _, _, _ ->
+    map(aa, bb, cc, dd, ee, ff)
+  }
+
+inline fun <A, B, C, D, E, F, G, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  f: NonEmptyList<F>,
+  g: NonEmptyList<G>,
+  map: (A, B, C, D, E, F, G) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d, e, f, g,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit
+  ) { aa, bb, cc, dd, ee, ff, gg, _, _, _ ->
+    map(aa, bb, cc, dd, ee, ff, gg)
+  }
+
+inline fun <A, B, C, D, E, F, G, H, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  f: NonEmptyList<F>,
+  g: NonEmptyList<G>,
+  h: NonEmptyList<H>,
+  map: (A, B, C, D, E, F, G, H) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d, e, f, g, h,
+    NonEmptyList.unit, NonEmptyList.unit
+  ) { aa, bb, cc, dd, ee, ff, gg, hh, _, _ ->
+    map(aa, bb, cc, dd, ee, ff, gg, hh)
+  }
+
+inline fun <A, B, C, D, E, F, G, H, I, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  f: NonEmptyList<F>,
+  g: NonEmptyList<G>,
+  h: NonEmptyList<H>,
+  i: NonEmptyList<I>,
+  map: (A, B, C, D, E, F, G, H, I) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d, e, f, g, h, i,
+    NonEmptyList.unit
+  ) { aa, bb, cc, dd, ee, ff, gg, hh, ii, _ ->
+    map(aa, bb, cc, dd, ee, ff, gg, hh, ii)
+  }
+
+inline fun <A, B, C, D, E, F, G, H, I, J, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  f: NonEmptyList<F>,
+  g: NonEmptyList<G>,
+  h: NonEmptyList<H>,
+  i: NonEmptyList<I>,
+  j: NonEmptyList<J>,
+  map: (A, B, C, D, E, F, G, H, I, J) -> Z
+): NonEmptyList<Z> =
+  NonEmptyList(
+    map(head, b.head, c.head, d.head, e.head, f.head, g.head, h.head, i.head, j.head),
+    tail.flatMap { aa ->
```

Comment:
I believe [this comment](https://github.com/arrow-kt/arrow/pull/2292#issuecomment-789809329) would apply here as well

Reply:
it does!
shall I fixed like done [here](https://github.com/arrow-kt/arrow/pull/2292/files#diff-69aba3b9460908255c0db2f50cc13506aada48c4e1a1d3d8f7bc9db8ff1f8616R104)?

Reply:
something like this:
```
inline fun <A, B, C, D, E, F, G, H, I, J, Z> NonEmptyList<A>.zip(
  b: NonEmptyList<B>,
  c: NonEmptyList<C>,
  d: NonEmptyList<D>,
  e: NonEmptyList<E>,
  f: NonEmptyList<F>,
  g: NonEmptyList<G>,
  h: NonEmptyList<H>,
  i: NonEmptyList<I>,
  j: NonEmptyList<J>,
  map: (A, B, C, D, E, F, G, H, I, J) -> Z
): NonEmptyList<Z> {
  val buffer = ArrayList<Z>()
  for (aa in tail) {
    for (bb in b.tail) {
      for (cc in c.tail) {
        for (dd in d.tail) {
          for (ee in e.tail) {
            for (ff in f.tail) {
              for (gg in g.tail) {
                for (hh in h.tail) {
                  for (ii in i.tail) {
                    for (jj in j.tail) {
                      buffer.add(map(aa, bb, cc, dd, ee, ff, gg, hh, ii, jj))
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
  return NonEmptyList(
    map(head, b.head, c.head, d.head, e.head, f.head, g.head, h.head, i.head, j.head),
    buffer
  )
}
```

Reply:
cc @nomisRev

Reply:
Actually we should:

I just tested and It seems the `unit` variable used for calling the higher arity function [has a bug](https://github.com/arrow-kt/arrow/blob/master/arrow-libs/core/arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt#L331) because it creates a `NonEmptyList(Unit, emptyList())` wich will do nothing for when we do `tail.flatMap` [here](https://github.com/arrow-kt/arrow/blob/master/arrow-libs/core/arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt#L435).

At some point we should add some tests as well

Reply:
What is the bug @aballano? If the `tail == emptyList` then there is nothing to `zip`/`mapN` there, right?

Reply:
@nomisRev Indeed, so for this:

    NonEmptyList.mapN(
        nonEmptyListOf(1, 2, 3),
        nonEmptyListOf(4, 5),
    ) { one, two ->
        println("->$one $two")
        ""
    }

there's just 1 iteration and prints "->1 4" and nothing else, because it just maps the head, not the tails.

Reply:
@nomisRev And BTW I know we are changing map to zip because of the stdlib, but we're doing it differently: the stdlib's zip combines 2 lists using the same index, meaning it does return a list with the shortest length of both:

listOf(1, 2).zip(listOf(4, 5, 6)) { one, two ->
        println("->$one $two")
    }

// ->1 4
// ->2 5

while our zip does a series of flatmaps and maps each value to each value:

// ->1 4
// ->1 5
// ->1 6
// ->2 4
// ->2 5
// ->2 6

Reply:
> there's just 1 iteration and prints "->1 4" and nothing else, because it just maps the head, not the tails.

Since `for(xx in emptyList<A>())` never triggers, right.

> but we're doing it differently: the stdlib's zip combines 2 lists using the same index, meaning it does return a list with the shortest length of bot

Right, `zip` is not a cartesian builder which is problematic for collections. 😐
I still think it's a better fit for `Either`, `Validated`, etc but that leaves open the question for cartesian builders for collections. Do we replace them, or remove them altogether.

I just ran into the same issue with `Map`, but there `mapN == zip` since `flatMap` works over the `union` of the `keys` but deriving with `emptyMap` is invalid.